drivers/can/can_mcan.c

```diff
@@ -752,13 +756,22 @@ int can_mcan_send(const struct can_mcan_config *cfg,
 	CACHE_CLEAN(&msg_ram->tx_buffer[put_idx].hdr, sizeof(tx_hdr));
 	CACHE_CLEAN(&msg_ram->tx_buffer[put_idx].data_32, ROUND_UP(data_length, 4));
 
-	data->tx_fin_cb[put_idx] = callback;
-	data->tx_fin_cb_arg[put_idx] = user_data;
+	if (!cfg->one_shot) {
+		data->tx_fin_cb[put_idx] = callback;
+		data->tx_fin_cb_arg[put_idx] = user_data;
+	}
```

Comment:
I think it makes sense to keep the callback.

Reply:
@alexanderwachter I agree. The callback will be called if the message is sent successfully. I've made the change to put the callback handling back the way it was.

This doesn't handle the case where the send fails, the callback won't ever fire, and `can_send()` will block.

Do you have any ideas about how to cleanly check the error condition, as in my comment https://github.com/zephyrproject-rtos/zephyr/pull/42152#issuecomment-1024517002 above?